<a href="https://colab.research.google.com/github/Saltizm/Data-mining-group-project/blob/main/pyspark_code.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [12]:
# init
import pyspark as s
from pyspark.ml.feature import StringIndexer, OneHotEncoder
from pyspark.ml.classification import RandomForestClassifier
from pyspark.sql import SparkSession
import pandas as pd;
import numpy as np;
import seaborn as sns;
import matplotlib.pyplot as plt;
import tqdm;
import os;
import sys;

In [13]:
# # mount
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [14]:
# Initialize SparkSession if not already initialized
spark = SparkSession.builder.appName("pyspark_code").getOrCreate()

In [15]:
if os.path.isdir(r'C:\Users\User\Desktop\Data-mining-group-project'):
    fpath = r'C:\Users\User\Desktop\Data-mining-group-project' #local
elif os.path.isdir(r'/content/drive/MyDrive/data mining'):
    fpath = r'/content/drive/MyDrive/data mining' #google drive
try:
    print(os.getcwd())
    train = pd.read_csv(fpath + os.sep + r'UNSW_NB15_training-set.csv')
    test = pd.read_csv(fpath + os.sep + r'UNSW_NB15_testing-set.csv')
except FileNotFoundError as e:
    print(f"error: {e}\nTry changing the training data directory in 'os.chdir'")

/content


In [ ]:
# STEP 1: Convert Pandas DataFrames to Spark DataFrames
print("Converting Pandas DataFrames to Spark DataFrames...")
spark_train_df = spark.createDataFrame(train)
spark_test_df = spark.createDataFrame(test)

print(f"Training DataFrame: {spark_train_df.count()} rows, {len(spark_train_df.columns)} columns")
print(f"Testing DataFrame: {spark_test_df.count()} rows, {len(spark_test_df.columns)} columns")
print("Spark DataFrames created successfully.")

# Task
Train and evaluate Random Forest, Logistic Regression, and XGBoost classifiers on the `spark_train_df` and `spark_test_df` datasets to predict the 'attack_cat' label, handling the unbalanced nature of the dataset. Provide a comprehensive summary of the evaluation results for each model and recommend the best-suited model based on performance metrics.

## Define Features (X) and Label (y)

### Subtask:
Clearly identify and separate the feature columns (X) and the target label column (y, which is 'attack_cat') from the existing `spark_train_df` and `spark_test_df`. Exclude non-feature columns like 'id'.


**Reasoning**:
The subtask requires identifying feature and label columns from the Spark DataFrames. First, I need to get all column names from one of the Spark DataFrames (as they should be consistent between train and test). Then, I will define the label column as 'attack_cat' and create a list of feature columns by excluding 'id', 'label', and 'attack_cat'.



In [16]:
columns = spark_train_df.columns
label_column_name = 'attack_cat'

# Exclude 'id', 'label', and the target label ('attack_cat') from feature columns
feature_columns = [col for col in columns if col not in ['id', 'label', label_column_name]]

print(f"Label column: {label_column_name}")
print(f"Number of feature columns: {len(feature_columns)}")
print(f"First 5 feature columns: {feature_columns[:5]}")

Label column: attack_cat
Number of feature columns: 42
First 5 feature columns: ['dur', 'proto', 'service', 'state', 'spkts']


## Build Feature Preprocessing Pipeline

### Subtask:
Construct a PySpark ML pipeline to preprocess the identified feature columns (X). This will involve applying `StringIndexer` and `OneHotEncoder` to categorical features, and then using `VectorAssembler` to combine all numerical and one-hot encoded features into a single 'features' vector column.


**Reasoning**:
This section implements a proper preprocessing pipeline that:
1. Identifies categorical vs numerical features using proper type checking
2. Creates a unified pipeline including feature preprocessing AND label indexing
3. Fits the pipeline ONCE on training data only (preventing data leakage)
4. Transforms both train and test using the same fitted pipeline
5. Uses handleInvalid='keep' everywhere to prevent row deletion
6. Verifies row counts to ensure no data loss

In [ ]:
from pyspark.ml.feature import StringIndexer, OneHotEncoder, VectorAssembler
from pyspark.ml import Pipeline
from pyspark.sql.types import StringType

# STEP 2: Identify categorical vs numerical features
categorical_columns = []
numerical_columns = []

for col_name in feature_columns:
    dtype = spark_train_df.schema[col_name].dataType
    if isinstance(dtype, StringType):
        categorical_columns.append(col_name)
    else:
        numerical_columns.append(col_name)

print(f"Categorical columns ({len(categorical_columns)}): {categorical_columns}")
print(f"Numerical columns ({len(numerical_columns)}): {numerical_columns[:5]}...")

# STEP 3: Build pipeline stages
# Create StringIndexer stages for categorical features
string_indexer_stages = [
    StringIndexer(inputCol=col, outputCol=col + '_indexed', handleInvalid='keep')
    for col in categorical_columns
]

# Create OneHotEncoder stages
one_hot_encoder_stages = [
    OneHotEncoder(inputCol=col + '_indexed', outputCol=col + '_encoded')
    for col in categorical_columns
]

# Prepare input columns for VectorAssembler
assembler_input_cols = numerical_columns + [
    col + '_encoded' for col in categorical_columns
]

# Create VectorAssembler stage (handleInvalid='keep' to prevent row deletion)
vector_assembler = VectorAssembler(
    inputCols=assembler_input_cols, 
    outputCol='features',
    handleInvalid='keep'
)

# Create label indexer stage (include in main pipeline)
label_indexer = StringIndexer(
    inputCol=label_column_name, 
    outputCol='indexed_label', 
    handleInvalid='keep'
)

# STEP 4: Combine all stages into a single Pipeline (includes label indexing)
all_stages = string_indexer_stages + one_hot_encoder_stages + [vector_assembler, label_indexer]
preprocessing_pipeline = Pipeline(stages=all_stages)

print(f"\nPreprocessing pipeline created with {len(all_stages)} stages")
print("Pipeline stages: StringIndexers -> OneHotEncoders -> VectorAssembler -> LabelIndexer")

# STEP 5: FIT pipeline ONCE on training data only
print("\nFitting pipeline on training data...")
fitted_pipeline = preprocessing_pipeline.fit(spark_train_df)
print("Pipeline fitted successfully.")

# STEP 6: Transform both train and test with the SAME fitted pipeline
print("\nTransforming training data...")
processed_train_df = fitted_pipeline.transform(spark_train_df).cache()
print("Transforming testing data...")
processed_test_df = fitted_pipeline.transform(spark_test_df).cache()
print("Transformations complete.")

# STEP 7: Verify no rows deleted
original_train_count = spark_train_df.count()
original_test_count = spark_test_df.count()
processed_train_count = processed_train_df.count()
processed_test_count = processed_test_df.count()

print(f"\nRow count verification:")
print(f"Training: {original_train_count} -> {processed_train_count} ({'✓ PASS' if original_train_count == processed_train_count else '✗ FAIL'})")
print(f"Testing: {original_test_count} -> {processed_test_count} ({'✓ PASS' if original_test_count == processed_test_count else '✗ FAIL'})")

assert processed_train_count == original_train_count, "Training rows were deleted!"
assert processed_test_count == original_test_count, "Testing rows were deleted!"
print("\n✓ All assertions passed - no rows deleted during preprocessing")

# Display schema
print("\nProcessed Training DataFrame Schema:")
processed_train_df.printSchema()

## Build Label Preprocessing Pipeline

### Subtask:
Create a simple PySpark ML pipeline to preprocess the target label column (y). This will involve using `StringIndexer` to convert the 'attack_cat' string labels into numerical indices, creating a 'label' column suitable for classification models. This also includes calculating class weights for unbalanced data.


**Reasoning**:
The subtask requires calculating class weights for the 'indexed_label' column to handle data imbalance. First, I need to get the counts for each unique 'indexed_label', the total number of samples, and the total number of unique classes from `processed_train_df`. Then, I will compute the class weights using the provided formula and store them in a new Spark DataFrame.



In [ ]:
from pyspark.sql.functions import col, lit

# STEP 8: Calculate class weights for imbalanced dataset
print("Calculating class weights...")

# 1. Calculate the class distribution of the 'indexed_label' column
class_counts = processed_train_df.groupBy('indexed_label').count()

# 2. Determine the total number of samples and the number of unique classes
total_samples = processed_train_df.count()
num_classes = class_counts.count()

# 3. Compute the class weight for each 'indexed_label'
# Formula: classWeight = total_samples / (num_classes * class_count)
class_weights_df = class_counts.withColumn(
    'classWeight',
    lit(total_samples) / (lit(num_classes) * col('count'))
).select('indexed_label', 'classWeight')

# Display calculated class weights
print("\nCalculated Class Weights:")
class_weights_df.show()

# 4. Join class weights with processed_train_df using LEFT JOIN to prevent row loss
print("\nJoining class weights with training data...")
count_before_join = processed_train_df.count()
processed_train_df = processed_train_df.join(class_weights_df, on='indexed_label', how='left').cache()
count_after_join = processed_train_df.count()

# STEP 9: Verify no rows deleted after join
print(f"Row count after join: {count_before_join} -> {count_after_join} ({'✓ PASS' if count_before_join == count_after_join else '✗ FAIL'})")
assert count_after_join == count_before_join, "Rows were deleted during class weights join!"
print("\n✓ Class weights added successfully with no row loss")

print(f"\nFinal Processed Training DataFrame Schema:")
processed_train_df.printSchema()

## Implement Random Forest Pipeline

### Subtask:
Develop a PySpark ML pipeline for the Random Forest Classifier. This pipeline will incorporate the preprocessed features and label, and will be configured to address the unbalanced nature of the dataset, potentially using class weights if the model supports it or other techniques.


**Reasoning**:
The subtask requires implementing a PySpark ML pipeline for the Random Forest Classifier, including setting the feature, label, and class weight columns.



In [21]:
from pyspark.ml.classification import RandomForestClassifier

# Instantiate RandomForestClassifier with specified parameters, including class weights
rf_classifier = RandomForestClassifier(
    featuresCol='features',
    labelCol='indexed_label',
    weightCol='classWeight',
    seed=42 # for reproducibility
)

# Create a Pipeline with the RandomForestClassifier
rf_pipeline = Pipeline(stages=[rf_classifier])

print("Random Forest Classifier pipeline created successfully.")

Random Forest Classifier pipeline created successfully.


## Implement Logistic Regression Pipeline

### Subtask:
Create a PySpark ML pipeline for the Logistic Regression Classifier, using the preprocessed features and label. Special attention will be paid to handling data imbalance, for example, by adjusting parameters like `weightCol` or exploring sampling techniques if necessary.


**Reasoning**:
The subtask requires implementing a PySpark ML pipeline for the Logistic Regression Classifier, including setting the feature, label, and class weight columns to handle data imbalance.



In [23]:
from pyspark.ml.classification import LogisticRegression

# Instantiate LogisticRegression with specified parameters, including class weights
lr_classifier = LogisticRegression(
    featuresCol='features',
    labelCol='indexed_label',
    weightCol='classWeight'
)

# Create a Pipeline with the LogisticRegression
lr_pipeline = Pipeline(stages=[lr_classifier])

print("Logistic Regression Classifier pipeline created successfully.")

Logistic Regression Classifier pipeline created successfully.


## Implement XGBoost Pipeline

### Subtask:
Set up a PySpark ML pipeline for the XGBoost Classifier, utilizing the preprocessed features and label. The model configuration will be tailored to mitigate the effects of the unbalanced dataset, such as setting appropriate scale_pos_weight or using sampling methods.


**Reasoning**:
The subtask requires implementing a PySpark ML pipeline for the XGBoost Classifier, including setting the feature, label, and class weight columns, and setting a seed for reproducibility.



In [25]:
from xgboost.spark import SparkXGBClassifier
from pyspark.ml import Pipeline

# Instantiate SparkXGBClassifier with specified parameters, including class weights
# Note: SparkXGBClassifier uses 'weight_col' for handling class imbalance, similar to other PySpark ML classifiers.
xgb_classifier = SparkXGBClassifier(
    features_col='features',
    label_col='indexed_label',
    weight_col='classWeight',
    seed=42 # for reproducibility
)

# Create a Pipeline with the SparkXGBClassifier
xgb_pipeline = Pipeline(stages=[xgb_classifier])

print("XGBoost Classifier pipeline created successfully.")

XGBoost Classifier pipeline created successfully.


## Train and Evaluate Models

### Subtask:
Train Random Forest, Logistic Regression, and XGBoost models on the preprocessed training data and evaluate their performance on the test data using metrics suitable for unbalanced datasets.


**Reasoning**:
First, I will train the Random Forest model using the `rf_pipeline` on the `processed_train_df` and then make predictions on the `processed_test_df` to prepare for evaluation.



In [26]:
print("Training Random Forest model...")
rf_model = rf_pipeline.fit(processed_train_df)
print("Random Forest model trained.")

print("Making predictions with Random Forest model...")
rf_predictions = rf_model.transform(processed_test_df)
print("Random Forest predictions generated.")

Training Random Forest model...
Random Forest model trained.
Making predictions with Random Forest model...
Random Forest predictions generated.


**Reasoning**:
Now, I will train the Logistic Regression model using the `lr_pipeline` on the `processed_train_df` and then make predictions on the `processed_test_df`. Subsequently, I will train the XGBoost model using the `xgb_pipeline` and make predictions on the `processed_test_df`, completing instruction #3 of the subtask.



In [27]:
print("Training Logistic Regression model...")
lr_model = lr_pipeline.fit(processed_train_df)
print("Logistic Regression model trained.")

print("Making predictions with Logistic Regression model...")
lr_predictions = lr_model.transform(processed_test_df)
print("Logistic Regression predictions generated.")

print("\nTraining XGBoost model...")
xgb_model = xgb_pipeline.fit(processed_train_df)
print("XGBoost model trained.")

print("Making predictions with XGBoost model...")
xgb_predictions = xgb_model.transform(processed_test_df)
print("XGBoost predictions generated.")

Training Logistic Regression model...
Logistic Regression model trained.
Making predictions with Logistic Regression model...
Logistic Regression predictions generated.

Training XGBoost model...


INFO:XGBoost-PySpark:Running xgboost-3.1.3 on 1 workers with
	booster params: {'objective': 'multi:softprob', 'device': 'cpu', 'seed': 42, 'num_class': 10, 'nthread': 1}
	train_call_kwargs_params: {'verbose_eval': True, 'num_boost_round': 100}
	dmatrix_kwargs: {'nthread': 1, 'missing': nan}
INFO:XGBoost-PySpark:Finished xgboost training!


XGBoost model trained.
Making predictions with XGBoost model...
XGBoost predictions generated.


**Reasoning**:
I will import the `MulticlassClassificationEvaluator`, initialize evaluators for 'f1', 'weightedPrecision', 'weightedRecall', and 'accuracy' with the correct label and prediction columns, and then calculate and print these metrics for the Random Forest, Logistic Regression, and XGBoost models, addressing instructions #4, #5, and #6 of the subtask.



**Reasoning**:
The preprocessing steps have been successfully re-executed, ensuring consistent feature dimensions and correctly calculated class weights. Now, I will re-train the Random Forest model using the `rf_pipeline` on the `processed_train_df` and then make predictions on the `processed_test_df`, as per instructions #1 and #2 of the subtask.



In [31]:
print("Training Random Forest model...")
rf_model = rf_pipeline.fit(processed_train_df)
print("Random Forest model trained.")

print("Making predictions with Random Forest model...")
rf_predictions = rf_model.transform(processed_test_df)
print("Random Forest predictions generated.")

Training Random Forest model...
Random Forest model trained.
Making predictions with Random Forest model...
Random Forest predictions generated.


**Reasoning**:
Now, I will re-train the Logistic Regression model using the `lr_pipeline` on the `processed_train_df` and then make predictions on the `processed_test_df`. Subsequently, I will re-train the XGBoost model using the `xgb_pipeline` and make predictions on the `processed_test_df`, re-completing instruction #3 of the subtask.



In [32]:
print("Training Logistic Regression model...")
lr_model = lr_pipeline.fit(processed_train_df)
print("Logistic Regression model trained.")

print("Making predictions with Logistic Regression model...")
lr_predictions = lr_model.transform(processed_test_df)
print("Logistic Regression predictions generated.")

print("\nTraining XGBoost model...")
xgb_model = xgb_pipeline.fit(processed_train_df)
print("XGBoost model trained.")

print("Making predictions with XGBoost model...")
xgb_predictions = xgb_model.transform(processed_test_df)
print("XGBoost predictions generated.")

Training Logistic Regression model...
Logistic Regression model trained.
Making predictions with Logistic Regression model...
Logistic Regression predictions generated.

Training XGBoost model...


INFO:XGBoost-PySpark:Running xgboost-3.1.3 on 1 workers with
	booster params: {'objective': 'multi:softprob', 'device': 'cpu', 'seed': 42, 'num_class': 10, 'nthread': 1}
	train_call_kwargs_params: {'verbose_eval': True, 'num_boost_round': 100}
	dmatrix_kwargs: {'nthread': 1, 'missing': nan}
INFO:XGBoost-PySpark:Finished xgboost training!


XGBoost model trained.
Making predictions with XGBoost model...
XGBoost predictions generated.


**Reasoning**:
The preprocessing and model training steps have been successfully re-executed. Now, I will proceed to evaluate the Random Forest, Logistic Regression, and XGBoost models using `MulticlassClassificationEvaluator` for 'f1', 'weightedPrecision', 'weightedRecall', and 'accuracy', addressing instructions #4, #5, and #6 of the subtask.



In [33]:
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

# Initialize evaluators for different metrics
evaluator_f1 = MulticlassClassificationEvaluator(labelCol='indexed_label', predictionCol='prediction', metricName='f1')
evaluator_weighted_precision = MulticlassClassificationEvaluator(labelCol='indexed_label', predictionCol='prediction', metricName='weightedPrecision')
evaluator_weighted_recall = MulticlassClassificationEvaluator(labelCol='indexed_label', predictionCol='prediction', metricName='weightedRecall')
evaluator_accuracy = MulticlassClassificationEvaluator(labelCol='indexed_label', predictionCol='prediction', metricName='accuracy')

models = {
    "Random Forest": rf_predictions,
    "Logistic Regression": lr_predictions,
    "XGBoost": xgb_predictions
}

results = {}

for model_name, predictions_df in models.items():
    print(f"\nEvaluating {model_name} Model...")
    f1_score = evaluator_f1.evaluate(predictions_df)
    weighted_precision = evaluator_weighted_precision.evaluate(predictions_df)
    weighted_recall = evaluator_weighted_recall.evaluate(predictions_df)
    accuracy = evaluator_accuracy.evaluate(predictions_df)

    results[model_name] = {
        "F1 Score": f1_score,
        "Weighted Precision": weighted_precision,
        "Weighted Recall": weighted_recall,
        "Accuracy": accuracy
    }

    print(f"  F1 Score: {f1_score:.4f}")
    print(f"  Weighted Precision: {weighted_precision:.4f}")
    print(f"  Weighted Recall: {weighted_recall:.4f}")
    print(f"  Accuracy: {accuracy:.4f}")

print("\nEvaluation Summary:")
for model_name, metrics in results.items():
    print(f"\n{model_name}:")
    for metric_name, value in metrics.items():
        print(f"  {metric_name}: {value:.4f}")


Evaluating Random Forest Model...
  F1 Score: 0.6613
  Weighted Precision: 0.8393
  Weighted Recall: 0.5943
  Accuracy: 0.5943

Evaluating Logistic Regression Model...
  F1 Score: 0.6794
  Weighted Precision: 0.8284
  Weighted Recall: 0.6162
  Accuracy: 0.6162

Evaluating XGBoost Model...
  F1 Score: 0.7382
  Weighted Precision: 0.8611
  Weighted Recall: 0.6845
  Accuracy: 0.6845

Evaluation Summary:

Random Forest:
  F1 Score: 0.6613
  Weighted Precision: 0.8393
  Weighted Recall: 0.5943
  Accuracy: 0.5943

Logistic Regression:
  F1 Score: 0.6794
  Weighted Precision: 0.8284
  Weighted Recall: 0.6162
  Accuracy: 0.6162

XGBoost:
  F1 Score: 0.7382
  Weighted Precision: 0.8611
  Weighted Recall: 0.6845
  Accuracy: 0.6845


## Evaluation Summary and Model Recommendation

Based on the evaluation results, here's a summary of the performance of each model:

### Random Forest Classifier
*   **F1 Score**: 0.6613
*   **Weighted Precision**: 0.8393
*   **Weighted Recall**: 0.5943
*   **Accuracy**: 0.5943

### Logistic Regression Classifier
*   **F1 Score**: 0.6794
*   **Weighted Precision**: 0.8284
*   **Weighted Recall**: 0.6162
*   **Accuracy**: 0.6162

### XGBoost Classifier
*   **F1 Score**: 0.7382
*   **Weighted Precision**: 0.8611
*   **Weighted Recall**: 0.6845
*   **Accuracy**: 0.6845

### Recommendation

Comparing the performance metrics across all three models, the **XGBoost Classifier** demonstrates superior performance in predicting the 'attack_cat' label, especially considering the unbalanced nature of the dataset. It achieved the highest scores across all evaluated metrics:

*   **Highest F1 Score**: 0.7382 (indicating a better balance between precision and recall)
*   **Highest Weighted Precision**: 0.8611
*   **Highest Weighted Recall**: 0.6845
*   **Highest Accuracy**: 0.6845

The F1 score is particularly important for unbalanced datasets as it provides a better measure of a model's performance than accuracy alone. Given its higher F1 score, XGBoost is better at identifying positive cases (various attack categories) without generating too many false positives. The weighted precision and recall also suggest that the XGBoost model handles different class distributions more effectively.

Therefore, the **XGBoost Classifier** is the best-suited model among the three for this task, offering the most robust and balanced performance for predicting 'attack_cat' in the given dataset.

## Final Task

### Subtask:
Provide a comprehensive summary of the evaluation results for Random Forest, Logistic Regression, and XGBoost classifiers. Based on the performance metrics, recommend the best-suited model for the given unbalanced dataset and problem.


## Summary:

### Q&A
Based on the performance metrics, the **XGBoost Classifier** is the best-suited model for the given unbalanced dataset and problem.

### Data Analysis Key Findings
*   **Feature and Label Preparation:**
    *   `42` feature columns were identified, excluding 'id', 'label', and 'attack_cat'.
    *   Categorical features (`proto`, `service`, `state`) were correctly identified and processed using `StringIndexer` and `OneHotEncoder`.
    *   All features were combined into a single `features` vector column using `VectorAssembler`.
    *   The `attack_cat` label was indexed into a numerical `indexed_label` column.
    *   Class weights were calculated for `10` unique classes in the training data (e.g., `indexed_label` 9.0 had a weight of `134.88`, while `indexed_label` 0.0 had `0.31`) and incorporated into the training dataset to address data imbalance.
*   **Model Performance Evaluation:**
    *   **Random Forest Classifier:** Achieved an F1 Score of `0.6613`, Weighted Precision of `0.8393`, Weighted Recall of `0.5943`, and Accuracy of `0.5943`.
    *   **Logistic Regression Classifier:** Achieved an F1 Score of `0.6794`, Weighted Precision of `0.8284`, Weighted Recall of `0.6162`, and Accuracy of `0.6162`.
    *   **XGBoost Classifier:** Demonstrated the highest performance across all metrics with an F1 Score of `0.7382`, Weighted Precision of `0.8611`, Weighted Recall of `0.6845`, and Accuracy of `0.6845`.
*   **Model Comparison:** The XGBoost Classifier consistently outperformed both Random Forest and Logistic Regression models across all evaluated metrics, particularly excelling in F1 Score, which is crucial for unbalanced datasets as it balances precision and recall.

### Insights or Next Steps
*   The superior F1 score of XGBoost ($0.7382$) highlights its effectiveness in handling the unbalanced dataset, providing a better balance between correctly identified positive attack categories and minimizing false positives compared to the other models.
*   Further hyperparameter tuning for the XGBoost Classifier could potentially lead to even better performance. Additionally, exploring ensemble methods or advanced techniques for handling class imbalance (e.g., SMOTE for PySpark) could be beneficial.
